### ISSUE TITLE
Consider bringing back GravitySofteningGas in the future?

### ISSUE THREAD

**Issue by freddavies** [MP-Gadget/MP-Gadget#646]
I was recently testing out the new MPCDF "viper" machine by running small-box simulations along the lines of [this paper](https://arxiv.org/abs/2312.06763), using the most up-to-date version of the code instead of my long-outdated fork. However, even at fairly modest resolution I noticed really strong DM-gas particle pairing, such that the baryon glass was almost entirely erased by z ~ 20 (for this test I'm using the default MP-GenIC grid+glass ICs, but starting at z = 199). I narrowed it down to the recent deprecation of the adaptive softening for gas particles, which I had been using previously.

Here is a thin slice of DM (black) and gas (blue) particles in a voidy region of an adiabatic simulation with 16 kpc/h resolution without adaptive softening,
![Screenshot 2024-07-17 at 11 07 20](https://github.com/user-attachments/assets/03a6a4c4-09b7-4e49-b3db-c6bb099c25e7)
(Note that many of the DM particles are hiding behind gas particles, if you look closely...)

And here is how it looks if I revert to a commit soon before #630 and set `GravitySofteningGas = 0`,
![Screenshot 2024-07-17 at 11 07 41](https://github.com/user-attachments/assets/bf02e6e4-7ead-4434-a63a-b901bc2b3c47)

I'm happy with this workaround for now, and might end up doing some selective merging into my fork to stay updated, but I thought I would mention it anyway. Since including the functionality otherwise slows things down a bit, maybe it could be re-implemented with a compiler flag instead?

**Comment by sbird** [comment]
I would definitely accept a PR to bring it back: the slow-down is not excessive. It is mostly just extra code complexity and I didn't know anyone was using it.

**Comment by sbird** [comment]
I prefer not to add compiler flags because they tend not to get tested properly.

**Comment by sbird** [comment]
Actually - I just read https://arxiv.org/pdf/2212.06851 and I think our adaptive softening implementation (in addition to suppressing structure formation) does not conserve energy. See their figure D1 and also https://arxiv.org/abs/astro-ph/0610872. Would be interesting to implement the tidal tensor softening from Gizmo. Would you be willing to test that if I did?

**Comment by nianyic7** [comment]
Hi Fred and Simeon, 
Recently for doing the non-cosmological runs, I had to get back different species softening and here is a relevant [commit](https://github.com/nianyic7/MP-Gadget/commit/22236efdd05c32aae427612302cf73f8003afb0b). It is a fairly small change and maybe you can take a look and see if it is useful? This is not the adaptive softening though, just a fixed softening for various particle types.

**Comment by freddavies** [comment]
> Actually - I just read https://arxiv.org/pdf/2212.06851 and I think our adaptive softening implementation (in addition to suppressing structure formation) does not conserve energy. See their figure D1 and also https://arxiv.org/abs/astro-ph/0610872. Would be interesting to implement the tidal tensor softening from Gizmo. Would you be willing to test that if I did?

This sounds very cool, and I would be happy to test it out. It looks like it has a much more robust theoretical basis than the traditional implementation, which always felt like a bit of a hack (particularly treating the gas and DM differently).


> Hi Fred and Simeon, Recently for doing the non-cosmological runs, I had to get back different species softening and here is a relevant [commit](https://github.com/nianyic7/MP-Gadget/commit/22236efdd05c32aae427612302cf73f8003afb0b). It is a fairly small change and maybe you can take a look and see if it is useful? This is not the adaptive softening though, just a fixed softening for various particle types.

Thanks Nianyi! I think in this specific case I need some amount of dynamic range in the softening between overdense and void regions, but some of this code may still be useful in case the gas and DM have to be treated separately with the Gizmo method.

**Comment by sbird** [comment]
Ok! Well I will put this on my _very short_ todo list and be in touch :)

**Comment by sbird** [comment]
> This sounds very cool, and I would be happy to test it out. It looks like it has a much more robust theoretical basis than the traditional implementation, which always felt like a bit of a hack (particularly treating the gas and DM differently).

Yes, it always bothered me too (that's really why I removed it). It is uncomfortable having the softening depend on a property of the hydro solver.

**Comment by sbird** [comment]
See also: https://arxiv.org/abs/2303.00639
and
https://arxiv.org/abs/2008.09124
